proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,520 @@
+diagnostics need to be emitted so that the user knows why the resource type is 
+invalid, or why the register type isn't correct. This spec defines the behavior
+of the compiler for whenever an invalid register type is given, or for when a valid
+register type is used to bind a resource type that for whatever reason is 
+incompatible with that register type.
+
+## Motivation
+
+There are several cases in DXC in which diagnostics that are related to
+invalid register types are out of date or invalid. For example, in the case of:
+
+`float b : register(u4);`
+
+an error will be emitted recommending the use of the 'b, c, or i' register
+type. However the 'i' register type is no longer in support, and the 'b' 
+register type is only reserved for resource types that are constant buffers.
+It is worth noting that there is an overloading of the register(...) keyword
+using the 'c' register type to indicate an offset into a constant buffer for
+numeric types only, as opposed to specifying a resource binding.
+Additionally, it is possible the user is unaware that this variable won't 
+actually be used as a resource, but the compiler doesn't communicate that 
+to the user. It would be great for any HLSL developer to immediately have
+answers to the following questions:
+For any valid variable type that needs to be bound as a resource,
+what is the set of valid register types for that variable,
+and what are all the possible diagnostics that can be emitted if the variable
+is given a register type that isn't in the set of valid register types?
+The design below aims to specify the answer to these questions.
+
+## Proposed solution
+The following table lists all the proposed diagnostic
+names and diagnostic messages that are relevant in this spec:
+
+| Diagnostic Name | Diagnostic message | Description |
+|-|-|-|
+| err_hlsl_mismatching_register_type_and_variable_type | "error: unsupported resource register binding '%select{t\|u\|b\|s}0' on variable of type '%1'"| Emitted if a variable type that isn't accompanied with a resource class attribute is bound with a standard register type (`t`, `u`, `b`, or `s`). |
+| err_hlsl_unsupported_register_type_and_variable_type | "error: register binding type '%0' not supported for variable of type '%1'" | Emitted if a variable type is bound using an unsupported register type, like binding a float with the 'x' register type. |
+| err_hlsl_mismatching_register_type_and_resource_type | "error: %select{srv\|uav\|cbv\|sampler}2 type '%0' requires register type '%select{t\|u\|b\|s}2', but register type '%1' was used." | Emitted if a known resource type is bound using a standard but mismatching register type, e.g., RWBuffer<int> getting bound with 's'.|
+| err_hlsl_unsupported_register_type_and_resource_type | "error: invalid register type '%0' used; expected 't', 'u', 'b', or 's'"| Emitted if an unsupported register type is used to bind a resource, like 'y' being used with RWBuffer<int>.|

```

Comment:
Is this up to date with the latest design?

Reply:
The actual implementation needs some updating given the new handling for when invalid registers are detected. Otherwise, I do believe the spec is self-consistent now.